mms_msg/sampling/pattern/meeting/state_based/transition_model.py

```diff
@@ -0,0 +1,315 @@
+        Initialize a Markov Model with n states. The states can be named with the states parameter.
+        Default naming is a simple enumeration.
+
+        Args:
+            probability_matrix: transition matrix for the markov model.
+                Must be a stochastic matrix of the dimensions n x n.
+            s0: index of the starting state or name of the starting state.
+                When s0 is an integer s0 it is interpreted as the index of the starting state.
+            state_names: (optional) names of the states, if used all states have to be named.
+                As name every type except integer can be used,
+                this type is reserved for the default enumeration.
+        """
+
+        # Assert correct dimensions and correct probabilities
+        if not probability_matrix.shape[0] > 0:
+            raise ValueError('The markov model requires at least one state')
+        if probability_matrix.shape[0] != probability_matrix.shape[1]:
+            raise ValueError(f'Probability matrix has wrong dimensions: Expecting a square matrix, '
+                             f'but got {probability_matrix.shape}.')
+        if not np.allclose(np.sum(probability_matrix, axis=1), np.ones(probability_matrix.shape[0]),
+                           rtol=0.0, atol=1e-06):
+            raise ValueError('The probabilities in some row, are not adding up to 1')
+
+        # Assert that when states are named, all states have names
+        if not (state_names is None or len(state_names) == probability_matrix.shape[0]):
+            raise ValueError('When the states parameter is used, there has to be names for all states.')
+
+        # Assert that no integers are used as names
+        if state_names is not None and type(state_names[0]) is int:
+            raise TypeError('Integers are not allowed as names for states')
+
+        self.s0 = s0
+        self._size = probability_matrix.shape[0]
+        self.probability_matrix = probability_matrix
+        if state_names is None:
+            self._state_names = [*range(self.size)]
+        else:
+            self._state_names = deepcopy(state_names)
+
+        if isinstance(s0, int):
+            self.current_state_index = s0
+        else:
+            self.current_state_index = self.state_names.index(s0)
+
+        self.last_state_index = None
+
+    def next(self, rng: np.random.random = np.random.default_rng()) -> Union[int, MST]:
+        """
+        Simulates a step in the markov model and return the state after the step
+
+        Args:
+            rng: The numpy rng that should be used, the rng should generate a number in the interval [0,1).
+                When not set a uniform rng is used.
+
+        Returns: Name of the state of the model after the step is executed
+        """
+
+        self.last_state_index = self.current_state_index
+
+        self.current_state_index = self.state_names.index(self.simulate_step(self.state_names[self.current_state_index],
```

Comment:
I would move `simulate_step` -> `_simulate_step` which returns the index and use it in `simulate_step` and `next`. It's weird to use `index` here because you know the index in the computation.